dev/io.openliberty.security.oidcclientcore.internal/src/io/openliberty/security/oidcclientcore/token/TokenResponseValidator.java

```diff
@@ -168,12 +170,39 @@ public JwtClaims validate(TokenResponse tokenResponse) throws TokenValidationExc
     /**
      * @param clientConfig
      * @return
+     * @throws OidcDiscoveryException
      */
-    private String getIssuer() {
+    private String getIssuer() throws OidcDiscoveryException {
+        String issuer = getIssuerFromProviderMetadata();
+        if (issuer != null) {
+            return issuer;
+        }
+        return getIssuerFromDiscoveryMetadata();
+    }
+
+    private String getIssuerFromProviderMetadata() {
+        OidcProviderMetadata providerMetadata = clientConfig.getProviderMetadata();
+        if (providerMetadata != null) {
+            String issuer = providerMetadata.getIssuer();
+            if (issuer != null && !issuer.isEmpty()) {
+                if (tc.isDebugEnabled()) {
+                    Tr.debug(tc, "Issuer found in the provider metadata: [" + issuer + "]");
+                }
+                return issuer;
+            }
+        }
+        return null;
+    }
+
+    private String getIssuerFromDiscoveryMetadata() throws OidcDiscoveryException {
         String issuer = null;
-        issuer = clientConfig.getProviderMetadata().getIssuer();
-        if ((issuer == null || issuer.isEmpty()) && getProviderDiscoveryMetadadata() != null) {
-            return ((String) this.discoveredProviderMetadata.get("issuer"));
+        JSONObject providerDiscoveryMetadata = getProviderDiscoveryMetadadata();
+        if (providerDiscoveryMetadata != null) {
+            issuer = (String) providerDiscoveryMetadata.get(OidcDiscoveryConstants.METADATA_KEY_ISSUER);
+        }
+        if (issuer == null || issuer.isEmpty()) {
+            String nlsMessage = Tr.formatMessage(tc, "DISCOVERY_METADATA_MISSING_VALUE", OidcDiscoveryConstants.METADATA_KEY_ISSUER);
+            throw new OidcDiscoveryException(clientConfig.getClientId(), clientConfig.getProviderURI(), nlsMessage);
```

Comment:
This also may not necessarily be an error. I believe the old OIDC client code will use the host value from the token endpoint URL as the issuer value if one isn't found in the configuration; we could consider doing something similar.

Reply:
the issuer is required in the jakarta spec and in the oidc discovery spec, so an error should be thrown here.

https://github.com/jakartaee/security/blob/master/api/src/main/java/jakarta/security/enterprise/authentication/mechanism/http/openid/OpenIdProviderMetadata.java#L87-L91

https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderMetadata